source/common/config/datasource.h

```diff
@@ -25,6 +32,108 @@ std::string read(const envoy::api::v2::core::DataSource& source, bool allow_empt
  */
 absl::optional<std::string> getPath(const envoy::api::v2::core::DataSource& source);
 
+/**
+ * Callback for async data source.
+ */
+using AsyncDataSourceCb = std::function<void(std::string)>;
+
+class LocalAsyncDataProvider;
+using LocalAsyncDataProviderPtr = std::unique_ptr<LocalAsyncDataProvider>;
+
+class RemoteAsyncDataProvider;
+using RemoteAsyncDataProviderPtr = std::unique_ptr<RemoteAsyncDataProvider>;
+
+class LocalAsyncDataProvider {
+public:
+  LocalAsyncDataProvider(Init::Manager& manager, const envoy::api::v2::core::DataSource& source,
+                         bool allow_empty, Api::Api& api, AsyncDataSourceCb&& callback)
+      : init_target_("LocalAsyncDataProvider", [this, &source, allow_empty, &api, callback]() {
+          callback(DataSource::read(source, allow_empty, api));
+          init_target_.ready();
+        }) {
+    manager.add(init_target_);
+  }
+
+  ~LocalAsyncDataProvider() { init_target_.ready(); }
+
+  /**
+   * Create async data provider for local data source.
+   * @param manager init manager for coordinating initialization
+   * @param source data source configuration
+   * @param allow_empty if the data is allowed empty
+   * @param api "Public" API that different components use to interact with the various system
+   * abstractions
+   * @param callback data provider callback
+   * @return data provider for local data source
+   */
+  static LocalAsyncDataProviderPtr create(Init::Manager& manager,
```

Comment:
Not sure what is the value of this static method

Reply:
I will remove them.